### PR TITLE
[WIP] HTTP framing implementation

### PR DESCRIPTION
**Pull request by stammw**
Relates to #22.

### CODE REVIEW

File: quinn-h3/src/frame.rs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    fn encode_header<T: BufMut>(&self, buf: &mut T) {
+        println!("len is {}", self.len());
+        varint::write(self.len() as u64, buf).unwrap();
+        buf.put_u8(self.ftype().0);
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DataFrame {
+    payload: Bytes,
```
Comment (stammw) [inline quinn-h3/src/frame.rs]: I'm not sure about using `Bytes` here for the interface with the user. Won't it create data copies ? I tried to use `<T> where T: AsRef<[u8]>` without success as I've seen in `quinn-proto::frame::Stream` encoding impl. I believe it's a function-call ergonomics matter as well ?

File: quinn-proto/src/frame.rs [hunk 2]
```diff
@@ -48,6 +48,7 @@ pub trait FrameStruct {
     const SIZE_BOUND: usize;
 }
 
+#[macro_export]
```
Comment (Ralith) [inline quinn-proto/src/frame.rs]: This macro currently hardcodes special handling of transport protocol STREAM frames in the `Display` impl, which external code probably doesn't want. In lieu of a more abstract macro, I'd recommend not exporting this, and instead including a similar but distinct one in new crate.

File: quinn-h3/src/frame.rs (hunk 1 above)
Comment (Ralith) [reply quinn-h3/src/frame.rs]: I think using `Bytes` here is fine. The fancy polymorphic `Stream` type in `-proto` is an implementation detail, and is not actually conferring any benefit at all right now. Buffering and flow control concerns, as well as the difficulty and expense of storing generic types in non-generic data structures, generally dictate that copying occurs anyway.

File: quinn-h3/src/frame.rs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        Ok(Type(buf.get_u8()))
+    }
+    fn encode<B: BufMut>(&self, buf: &mut B) {
+        buf.write(self.0);
+    }
+}
+
+trait FrameHeader {
+    fn len(&self) -> usize;
+    fn ftype(&self) -> Type;
```
Comment (Ralith) [inline quinn-h3/src/frame.rs]: Could this be
```suggestion
    const TYPE: Type;
```
instead?

File: quinn-h3/src/frame.rs [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+impl HttpFrame {
+    pub fn encode<T: BufMut>(&self, buf: &mut T) {
+        match self {
+            HttpFrame::Data(f) => f.encode(buf),
+            HttpFrame::Settings(f) => f.encode(buf),
+        }
+    }
+
+    pub fn decode<T: Buf>(buf: &mut T) -> Result<Self, Error> {
+        let len = varint::read(buf).ok_or(Error::Malformed)?;
```
Comment (Ralith) [inline quinn-h3/src/frame.rs]: `BufExt` defines `fn get_var(&mut self) -> coding::Result<u64>`, which would be more concise and preclude needing to publicize/import `proto::varint`.

File: quinn-h3/src/frame.rs [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+    fn decode<T: Buf>(buf: &mut T) -> Result<SettingsFrame, Error> {
+        let mut settings = SettingsFrame::default();
+        while buf.has_remaining() {
+            if buf.remaining() < 3 {
+                // remains less than id + minimum-size varint
+                return Err(Error::Malformed);
+            }
+            let identifier = buf.get::<SettingId>()?;
+            let value = varint::read(buf).ok_or(Error::Malformed)?;
```
Comment (Ralith) [inline quinn-h3/src/frame.rs]: As above.

File: quinn-h3/src/frame.rs [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        buf.write(self.0);
+    }
+}
+
+trait FrameHeader {
+    fn len(&self) -> usize;
+    fn ftype(&self) -> Type;
+    fn encode_header<T: BufMut>(&self, buf: &mut T) {
+        println!("len is {}", self.len());
+        varint::write(self.len() as u64, buf).unwrap();
```
Comment (Ralith) [inline quinn-h3/src/frame.rs]: `BufMutExt` defines a `write_var` helper that may be of interest, in this and other cases.

File: quinn-proto/src/frame.rs (hunk 2 above)
Comment (stammw) [reply quinn-proto/src/frame.rs]: Yes, this module doesn't actually need to have use the same macro. Fixed.

File: quinn-h3/src/frame.rs (hunk 3 above)
Comment (stammw) [reply quinn-h3/src/frame.rs]: Oh, didn't know about associated const. Rust is neat.

**Comment by stammw** [comment]
Almost done !

Still have to implement ignoring some frame fromated as described in `4.2.10. Reserved Frame Types ` and a last check against the specs.

**Comment by stammw** [comment]
Changes imported in #136

**Pull request closed without merging**